air/src/execution_step/lambda_applier/applier.rs

```diff
@@ -132,31 +157,74 @@ fn select_by_path_from_canon_map<'value>(
     // There will be an empty canon stream if the key was not found.
     let result = match (NonEmpty::try_from(body.to_vec()), canon_stream) {
         (Ok(body_part), Some(canon_stream)) => {
-            let canon_stream_iter = canon_stream.iter().map(|v| v.get_result().deref());
-            let LambdaResult { result, .. } = select_by_path_from_stream(canon_stream_iter, &body_part, exec_ctx)?;
-            result
+            // csm.$.key... case
+
+            let canon_stream_iter = canon_stream.iter().map(|v| (v.get_result().deref(), v.get_tetraplet()));
+            select_by_path_from_canon_map_stream(canon_stream_iter, &body_part, exec_ctx)?
         }
         (Err(..), Some(canon_stream)) => {
-            let value = canon_stream.as_jvalue();
-            Cow::Owned(value)
+            // csm.$.key case
+            let prefix_with_path = false;
+            let tetraplet = update_tetraplet_with_path(canon_map.tetraplet(), original_lambda, prefix_with_path);
+            let value = Cow::Owned(canon_stream.as_jvalue());
+
+            MapLensResult::from_cow(value, tetraplet)
         }
         _ => {
-            let SecurityTetraplet {
-                peer_pk,
-                service_id,
-                function_name,
-                json_path,
-            } = canon_map.tetraplet().as_ref();
-            let json_path = json_path.to_string() + &prefix.to_string();
-
-            let tetraplet = SecurityTetraplet::new(peer_pk, service_id, function_name, json_path).into();
-            let value = CanonStream::new(vec![], tetraplet).as_jvalue();
-            Cow::Owned(value)
+            // csm.$.non_existing_key case
+            let prefix_with_path = false;
+            let tetraplet = update_tetraplet_with_path(canon_map.tetraplet(), original_lambda, prefix_with_path);
```

Comment:
`update_tetraplet_with_path` should be simplified and split into two functions: `update_...` witch can use `SecurityTetraplet::add_path`, and `replace_...` that may use struct update.  The flag seem to make code less clear.